library/core/src/mem/maybe_uninit.rs

```diff
@@ -1061,8 +1061,8 @@ impl<T> MaybeUninit<T> {
     where
         T: Copy,
     {
-        // SAFETY: &[T] and &[MaybeUninit<T>] have the same layout
-        let uninit_src: &[MaybeUninit<T>] = unsafe { super::transmute(src) };
+        // SAFETY: `[T]` and `[MaybeUninit<T>]` have the same layout.
+        let uninit_src = unsafe { &*(src as *const [T] as *const [MaybeUninit<T>]) };
```

Comment:
It might make sense to use `ptr::from_ref` and `.cast::<T>()` instead of `as` casts for these right away, see https://github.com/rust-lang/rust/pull/109255.

Reply:
```suggestion
        let uninit_src = unsafe { &*ptr::from_ref(src).cast::<[MaybeUninit<T>]>() };
```

Reply:
I agree with this change, but I'd rather land it separately. IMO, if we decide to change these pointer casts (which are everywhere in core) they should be changed all at once, preferably with a lint.

Reply:
this is introducing new uses of `as`, which can be avoided by using those methods.